api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
 package aws
 
-import "fmt"
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"sort"
+	"strconv"
+
+	"github.com/zpatrick/go-bytesize"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/api/job"
+	"github.com/quintilesims/layer0/api/provider"
+	awsc "github.com/quintilesims/layer0/common/aws"
+	"github.com/quintilesims/layer0/common/config"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]models.ResourceConsumer
```

Comment:
Not thread safe (or should that be go routine safe). Not that scale can be called concurrently via the JobRunner and the Scaler timer. Maybe a RWMutex or a concurrent collection?

Reply:
Hmm maybe use this package? https://github.com/orcaman/concurrent-map

Reply:
or possible [go-cache](https://github.com/zpatrick/go-cache) :trollface: